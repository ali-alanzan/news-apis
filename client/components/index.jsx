import { useEffect, useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import { toast } from "react-toastify";
import '../css/reset.css';
import { fetchJSON } from './utils'
import { useParams } from 'react-router-dom';


const colorMain = '#6a59ca';


export function Header({account}) {
  const acc = account && account.name ? account : false
  
  return (
    <div style={{
      width: '100%',
      backgroundColor: colorMain,
      color: '#e2e2e2',
      height: '55px',
      display: 'flex',
      justifyContent: 'space-between',
      padding: '.5rem 4rem'
    }}>

      <div style={{flexGrow: 1}} className="logo-container">
        <h1>News App 
        
          {!acc ? '':<><button><a href="/add">Add</a></button><button><a href="/myarticles">My articles</a></button></>}

          
        </h1>
      </div>



      <div style={{}}>
        {!acc ? <Link to="/login">Login</Link>:acc.name}
        
      </div>
      

    </div>
  );
}


export function Sidebar() {
  const [articles, setArticles] = useState([]);
  useEffect(async () => {
    await fetch('/api/news').then(res => res.json().then(data => setArticles(data)));      
  }, []);
  return (
    <div style={{
      width: '30%',
      backgroundColor: colorMain,
      height: '100%',
      minHeight: '100vh',
      display: 'flex',
      flexDirection: 'column',
      padding: '.5rem 4rem',
      float: 'left'
    }}>  
    


      <div className="topics-container">
        <div className="d-f">
          <a href="#1">
            Topic(123)
          </a>

          <a href="#1">
            Topic(1234)
          </a>

          <a href="#1">
            Topic(124)
          </a>

          <a href="#1">
            Topic(14)
          </a>
        </div>
      </div>



      <div className="d-f articlessidebar-container">
      {articles.length>0 && articles.map((article) => (
          <a  key={`${article._id}`} href={`/edit/${article._id}`}>{article.title}</a>
      ))}
      </div>
      

    </div>
  );
}


export function FrontPage() {
  const [articles, setArticles] = useState([]);
  useEffect(async () => {
    await fetchJSON('/api/news').then(res => setArticles(res));      

  }, [articles]);

  return (
    <div className="main">
      <div className="d-f articlessidebar-container">
      {articles.length>0 && articles.map((article) => (
        <div key={`${article._id}`}>
          <a href={`/edit/${article._id}`}>{article.title}</a>
        </div>
      ))}
       
      </div>
    </div>
  );
}



export function SingleArticle () {
  return (
    <div style={{
      width: '68%',
      float: 'left',
      margin: '2% 0 0 2%'
    }}>
      <div className="d-f articlessidebar-container">
        <h1>Title</h1>
        <span>Date mm/yy/dd</span>
        
        <div className="topics-container">
          <div className="d-f">
            <a href="#1">
              Topic(123)
            </a>

            <a href="#1">
              Topic(1234)
            </a>

            <a href="#1">
              Topic(124)
            </a>

            <a href="#1">
              Topic(14)
            </a>
          </div>
        </div>
        
        <p>
          Article text
        </p>
      </div>
    </div>
  )
}



export function AddArticle ({account}) {
  if(account.email == undefined) {
    return <h1>Please Login</h1>
  }
  const navigate = useNavigate();
  const [values, setValues] = useState({}) 
  const onChange = (e) => {
    setValues({...values, [e.target.name]: e.target.value })
  }
  const onAddArticle = async (e) => {
    e.preventDefault();
    const res = await fetch("/api/news/add", {
      method: "POST",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify({...values, author: account.email}),
    });
    if (res.ok) {
      toast.success('Your article added successfully');
      navigate('/myarticles')

    } else {
      setError(`Failed ${res.status} ${res.statusText}`);
    }
  }
  return (
    <form onSubmit={onAddArticle}>
      <div>
        <label>Title</label>
        <input name="title" onChange={onChange} required/>
      </div>

      <div>
        <label>Text</label>
        <textarea name="text" onChange={onChange} cols="30" rows="4" required/>
      </div>

      <div>
        <label>Category</label>
        <select name="category" onChange={onChange} required>
          <option value="">--choose--</option>
          <option value="Health">Health</option>
          <option value="Technology">Technology</option>
          <option value="Global">Global</option>
        </select>
      </div>

      <div>
        <input type="submit" value="Add" />
      </div>
    </form>
  )
}


export function EditArticle({account}) {
  if(account.email == undefined) {
    return <h1>Please Login</h1>
  }
  const [values, setValues] = useState({title: '', text: '', category: ''})
  const { slug } = useParams();

  useEffect(async () => {
    await fetch(`/api/news/${slug}`).then(res => res.json().then(data => setValues(data)));      
  }, []);
  const navigate = useNavigate();
  
  const onChange = (e) => {
    setValues({...values, [e.target.name]: e.target.value })
  }
  const onAddArticle = async (e) => {
    e.preventDefault();
    const res = await fetch("/api/news/save", {
      method: "POST",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify({...values, author: account.email}),
    });
    if (res.ok) {
      toast.success('Your article added successfully');
      navigate('/myarticles')

    } else {
      setError(`Failed ${res.status} ${res.statusText}`);
    }
  }
  return (
    <form onSubmit={onAddArticle}>
      <div>
        <label>Title</label>
        <input name="title" onChange={onChange} value={values.title} required/>
      </div>

      <div>
        <label>Text</label>
        <textarea name="text" onChange={onChange} cols="30" rows="4" required defaultValue={values.text} />
      </div>

      <div>
        <label>Category</label>
        <select name="category" onChange={onChange} value={values.category} required>
          <option value="">--choose--</option>
          <option value="Health">Health</option>
          <option value="Technology">Technology</option>
          <option value="Global">Global</option>
        </select>
      </div>

      <div>
        <input type="submit" value="Add" />
      </div>
    </form>
  )
}


export function MyArticles ({account}) {
  const [articles, setArticles] = useState([]);
  useEffect(async () => {
    if(account.email!=undefined&&articles.length<=0) {
      await fetch('/api/news/?author='+account.email).then(res => res.json()).then(res => setArticles(res));      
    }
    
  }, [account]);

  if(account.email==undefined) { return <p>Please login</p> }
  return (
    <div className="main">
      <h1>My Articles</h1>
      {articles.length>0 && articles.map((article) => (
        <div key={`${article._id}`}>
          <a href={`/view/${article.slug}`}>{article.title}</a>
          
          <a href={`/edit/${article.slug}`}><button>Edit</button></a>
        </div>
      ))}
    </div>
  )

}