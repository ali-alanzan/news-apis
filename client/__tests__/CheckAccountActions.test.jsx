
import React from "react";
import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";
import { Header,  } from "../components";

const origin =  window.location.origin;
const readerAccount = {email: 'email@email.em', name:'John', google: true}
const writerAccount = {email: 'email@email.em', name:'John'}
describe("Account Action in Header", () => {


    it("shows Login buttons(Google, Azure)", async () => {

        const domElement = document.createElement("div");
        await act(async () => {
          ReactDOM.render(<Header />, domElement);
        });
        expect(
            Array.from(domElement.querySelectorAll("a")).map((e) => e.innerHTML)
          ).toEqual(["News App", "Writer Login", "Join by Google"]);

        expect(
        Array.from(domElement.querySelectorAll("a")).map((e) => e.href)
        ).toEqual([origin+"/", origin+"/login",origin+"/logingoogle"]);


        expect(domElement.innerHTML).toMatchSnapshot();
    });  
    
    
    it("shows Logout", async () => {

        const domElement = document.createElement("div");
        await act(async () => {
          ReactDOM.render(<Header account={readerAccount} />, domElement);
        });

        expect(domElement.querySelector("#logout").innerHTML).toEqual(
            "Logout"
          );

          expect(domElement.querySelector("#logout").href).toEqual(
            origin+"/Logout"
          );

        expect(domElement.innerHTML).toMatchSnapshot();
    });  


    it("shows ADD And MyArticles And Logout Buttons for writer", async () => {

        const domElement = document.createElement("div");
        await act(async () => {
          ReactDOM.render(<Header account={writerAccount} />, domElement);
        });
        expect(
            Array.from(domElement.querySelectorAll("div a")).map((e) => e.innerHTML)
          ).toEqual(["News App", "Add", "My articles", "Logout"]);

        expect(
        Array.from(domElement.querySelectorAll("div a")).map((e) => e.href)
        ).toEqual([origin+"/", origin+"/add", origin+"/myarticles", origin+"/Logout"]);


        expect(domElement.innerHTML).toMatchSnapshot();
    });
    

});