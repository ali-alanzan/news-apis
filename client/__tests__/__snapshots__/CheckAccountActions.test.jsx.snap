// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account Action in Header shows ADD And MyArticles And Logout Buttons for writer 1`] = `"<div style=\\"width: 100%; background-color: rgb(106, 89, 202); color: rgb(226, 226, 226); height: 55px; display: flex; justify-content: space-between; padding: .5rem 4rem;\\"><div style=\\"flex-grow: 1;\\" class=\\"logo-container\\"><h1><a href=\\"/\\">News App</a><button><a href=\\"/add\\">Add</a></button><button><a href=\\"/myarticles\\">My articles</a></button></h1></div><div>John<a href=\\"/Logout\\" id=\\"logout\\" style=\\"margin: 0px 10px; color: rgb(0, 0, 0);\\">Logout</a></div></div>"`;

exports[`Account Action in Header shows Login buttons(Google, Azure) 1`] = `"<div style=\\"width: 100%; background-color: rgb(106, 89, 202); color: rgb(226, 226, 226); height: 55px; display: flex; justify-content: space-between; padding: .5rem 4rem;\\"><div style=\\"flex-grow: 1;\\" class=\\"logo-container\\"><h1><a href=\\"/\\">News App</a></h1></div><div><div class=\\"userarea\\"><a href=\\"/login\\">Writer Login</a><a href=\\"/logingoogle\\" style=\\"color: yellow;\\">Join by Google</a></div></div></div>"`;

exports[`Account Action in Header shows Logout 1`] = `"<div style=\\"width: 100%; background-color: rgb(106, 89, 202); color: rgb(226, 226, 226); height: 55px; display: flex; justify-content: space-between; padding: .5rem 4rem;\\"><div style=\\"flex-grow: 1;\\" class=\\"logo-container\\"><h1><a href=\\"/\\">News App</a></h1></div><div>John<a href=\\"/Logout\\" id=\\"logout\\" style=\\"margin: 0px 10px; color: rgb(0, 0, 0);\\">Logout</a></div></div>"`;
